infrastructure/src/main/java/org/corfudb/infrastructure/log/statetransfer/StateTransferManager.java

```diff
@@ -0,0 +1,307 @@
+                                        return synchronousStateTransfer(batchProcessor, batchStream)
+                                                .thenApply(result ->
+                                                        createStatusBasedOnTransferResult(result,
+                                                                numAddressesToTransfer));
+
+                                    }
+                                    // If a segment contains some other status -> return.
+                                } else {
+                                    return CompletableFuture.completedFuture(status);
+                                }
+                            });
+                    return CurrentTransferSegment
+                            .builder()
+                            .startAddress(segment.getStartAddress())
+                            .endAddress(segment.getEndAddress())
+                            .status(newStatus)
+                            .build();
+                }
+
+
+        ).collect(ImmutableList.toImmutableList());
+    }
+
+    /**
+     * Given a batch processor and a stream of batches, execute the state transfer.
+     *
+     * @param batchProcessor An instance of a batch processor.
+     * @param batchStream    A stream of batches.
+     * @return A completed future containing a result of total number of addresses transferred
+     * or an exception.
+     */
+    CompletableFuture<Result<Long, StreamProcessFailure>> synchronousStateTransfer(
+            StateTransferBatchProcessor batchProcessor, Stream<Batch> batchStream) {
+        Result<Long, StreamProcessFailure> accumulatedResult = Result.ok(0L);
+
+        Result<Long, StreamProcessFailure> resultOfTransfer =
+                batchStream.reduce(accumulatedResult, (resultSoFar, nextBatch) -> {
+                    BatchResult batchResult = CFUtils
+                            .getUninterruptibly(batchProcessor.transfer(nextBatch));
+                    if (batchResult.getStatus() == BatchResult.FailureStatus.FAILED) {
+                        return Result.error(new StreamProcessFailure());
+                    } else {
+                        return resultSoFar
+                                .map(sumSoFar -> sumSoFar + batchResult.getAddresses().size());
+                    }
+                }, (oldSum, newSum) -> newSum);
+
+        return CompletableFuture.completedFuture(resultOfTransfer);
+    }
+
+
+    /**
+     * Based on a result of a state transfer, update a segment status.
+     *
+     * @param result      A result, containing a total number of records transferred or a failure.
+     * @param totalNeeded A total number of records that must be transferred.
+     * @return An updated segment status.
+     */
+    CurrentTransferSegmentStatus createStatusBasedOnTransferResult(Result<Long,
+            StreamProcessFailure> result, long totalNeeded) {
+        Result<Long, StreamProcessFailure> checkedResult =
```

Comment:
please reformat this block of code. It's hard to read

Reply:
Done